### PR TITLE
Plugins: Remove connection/hop-by-hop request/response headers for call resource

### LINKED ISSUE #60076: Plugins: Remove hop-by-hop request/response headers for call resource
Call resource should remove request/response connection/hop headers in similar manner as Go's reverse proxy functions:
- https://cs.opensource.google/go/go/+/refs/tags/go1.19.4:src/net/http/httputil/reverseproxy.go;l=274-281
- https://cs.opensource.google/go/go/+/refs/tags/go1.19.4:src/net/http/httputil/reverseproxy.go;l=328-332

In addition, prometheus deletes "stop" and hop headers for call resource:
https://github.com/grafana/grafana/blob/3c9ccd45121aeef5cac4ed0e5f1718bb33c436f1/pkg/tsdb/prometheus/resource/resource.go#L71-L72

### PR DESCRIPTION
**Pull request by marefr**
**What is this feature?**
Removes request/response connection/hop headers for call resource in similar manner as Go's reverse proxy functions. Also removes Prometheus datasource custom call resource header manipulation in regards to hop-by-hop headers.

**Why do we need this feature?**
Call resource should remove request/response connection/hop headers in similar manner as Go's reverse proxy functions:
- https://github.com/golang/go/blob/dc04f3ba1f25313bc9c97e728620206c235db9ee/src/net/http/httputil/reverseproxy.go#L274-L281
- https://github.com/golang/go/blob/dc04f3ba1f25313bc9c97e728620206c235db9ee/src/net/http/httputil/reverseproxy.go#L328-L332 

In addition, prometheus deletes "stop" and hop headers for call resource and shouldn't be needed any longer after these changes:
https://github.com/grafana/grafana/blob/3c9ccd45121aeef5cac4ed0e5f1718bb33c436f1/pkg/tsdb/prometheus/resource/resource.go#L71-L72 

**Who is this feature for?**
Plugin platform/developers.

**Which issue(s) does this PR fix?**:
Fixes #60076 
Ref #58646 

**Special notes for your reviewer**:
Prometheus call resource header manipulation removed forwarding of Cookie headers. This is now removed and should be automatically handled by https://github.com/grafana/grafana/blob/main/pkg/services/pluginsintegration/clientmiddleware/cookies_middleware.go

### CODE REVIEW

File: pkg/plugins/manager/client/client.go [hunk 1]
```diff
@@ -204,3 +218,73 @@ func (s *Service) plugin(ctx context.Context, pluginID string) (*plugins.Plugin,
...
+				delHeader = k
+				break
+			}
+		}
+
+		if delHeader != "" {
+			delete(h, delHeader)
+		}
+	}
+}
```
Comment (wbrowne) [inline pkg/plugins/manager/client/client.go:284]: I think it'd be simpler to do: 

```suggestion
func removeHopByHopHeaders(h map[string][]string) {
	for _, hh := range hopHeaders {
		for k := range h {
			if hh == textproto.CanonicalMIMEHeaderKey(k) {
				delete(h, k)
				break
			}
		}
	}
}
```

Same situation for the connection headers

File: pkg/plugins/manager/client/client_test.go [hunk 2]
```diff
@@ -71,8 +72,142 @@ func TestQueryData(t *testing.T) {
...
 
+func TestCallResource(t *testing.T) {
+	registry := fakes.NewFakePluginRegistry()
+	p := &plugins.Plugin{
+		JSONData: plugins.JSONData{
+			ID: "pid",
+		},
+	}
+
+	const backendResponse = "I am the backend"
```
Comment (wbrowne) [inline pkg/plugins/manager/client/client_test.go:83]: 😂  

Walt

File: pkg/plugins/manager/client/client.go [hunk 3]
```diff
@@ -204,3 +218,73 @@ func (s *Service) plugin(ctx context.Context, pluginID string) (*plugins.Plugin,
 
 	return p, true
 }
+
+// removeConnectionHeaders removes hop-by-hop headers listed in the "Connection" header of h.
+// See RFC 7230, section 6.1
+//
+// Based on https://cs.opensource.google/go/go/+/refs/tags/go1.19.4:src/net/http/httputil/reverseproxy.go;l=411-421
```
Comment (itsmylife) [inline pkg/plugins/manager/client/client.go:225]: I cannot open this link. It says permission denied. 
I think the link must be one of the followings:
- https://go.googlesource.com/go/+/refs/tags/go1.19.4/src/net/http/httputil/reverseproxy.go
- https://github.com/golang/go/blob/release-branch.go1.19/src/net/http/httputil/reverseproxy.go

**Review by itsmylife — APPROVED** [review]
LGTM

**Review by xnyo — APPROVED** [review]
LGTM! :+1:

File: pkg/plugins/manager/client/client.go (hunk 1 above)
Comment (marefr) [reply pkg/plugins/manager/client/client.go:284]: Hmm is it really a good advice to delete a key in a map while you're iterating it?

Comment (marefr) [reply pkg/plugins/manager/client/client.go:284]: Oh seems like it is. I'll change

Comment (wbrowne) [reply pkg/plugins/manager/client/client.go]: Yeah good old Golang!

**Review by wbrowne — APPROVED** [review]
LGTM

**Pull request merged**